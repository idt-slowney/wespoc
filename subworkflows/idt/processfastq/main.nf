include { FASTQC as PreprocessFASTQC }  from '../../../modules/idt/fastqc/main'
include { SEQTK_SAMPLE }                from '../../../modules/idt/seqtk/sample/main'
include { CUTADAPT_EXTRACTUMI }         from '../../../modules/idt/cutadapt/extractumi/main'
include { TRIMGALORE }                  from '../../../modules/idt/trimgalore/main'
include { FASTQC as PostprocessFASTQC } from '../../../modules/idt/fastqc/main'

workflow PROCESSFASTQ {

    take:
    reads               // channel: [ meta, reads ]
    skip_fastqc         // bool: [mandatory] default = false
    skip_downsample     // bool: [mandatory] default = false
    umi_flag            // bool: [mandatory] default = false
    skip_extractumi     // bool: [mandatory] default = true
    read_structure      // str: [optional] default = '8M143T 8M143T', aligns with PRISM UMI kit
    skip_trimming       // bool: [mandatory] default = false

    main:
    // Instantiate empty channels
    ch_versions = Channel.empty()
    fastqc_html = Channel.empty()
    fastqc_zip = Channel.empty()

    // Preprocessing QC -- skip_fastqc (DEFAULT: FALSE)
    if (!skip_fastqc) {
        PreprocessFASTQC(reads)
        ch_versions = ch_versions.mix(PreprocessFASTQC.out.versions)
        fastqc_html = fastqc_html.mix(PreprocessFASTQC.out.html)
        fastqc_zip = fastqc_zip.mix(PreprocessFASTQC.out.zip)
    }

    // Downsampling -- skip_downsample (DEFAULT: FALSE)
    // For now - we branch and mix here to account for situations with partial downsampling
    // TODO - decide if we want to keep it like this, where downsampling is done per sample in the samplesheet
    // or set one downsampling level as a parameter for all samples
    sample_reads = reads
    if (!skip_downsample) {
        downsample_reads = sample_reads.branch{
                    Sample: it[0]["sample_size"] && it[0]["sample_size"] != null
                    No_Sample: it[0]["sample_size"] == null | !("sample_size" in it[0]) }
                    
        SEQTK_SAMPLE(
            downsample_reads.Sample
            .map{ meta, reads -> [meta, reads, meta.sample_size] } )
        ch_versions = ch_versions.mix(SEQTK_SAMPLE.out.versions)
        sample_reads = downsample_reads.No_Sample.mix(SEQTK_SAMPLE.out.reads)
    }

    // ExtractUMI -- umi_flag (FALSE), skip_extractumi (FALSE)
    extractumi_reads = sample_reads
    if (umi_flag && !skip_extractumi) {
        CUTADAPT_EXTRACTUMI(extractumi_reads, read_structure)
        ch_versions = ch_versions.mix(CUTADAPT_EXTRACTUMI.out.versions)
        extractumi_reads = CUTADAPT_EXTRACTUMI.out.reads
    }

    // Trim Adapters -- skip_trimming (FALSE)
    trim_reads = extractumi_reads
    trimgalore_log          = Channel.empty()
    trimgalore_unpaired     = Channel.empty()
    trimgalore_html         = Channel.empty()
    trimgalore_zip          = Channel.empty()
    if (!skip_trimming) {
        TRIMGALORE(trim_reads)
        // Here we can include other trimming tools and control which get run with the when directive
        // i.e. TRIMMOMATIC(trim_reads) 
        // ---  { ext.when = trimmer = trimmomatic }
        // Then we mix results of all tools with a previously instantiated empty channel and yield only the output from the process that was run
        // i.e. trimmed_out = channel.empty()
        // ---  trimmed_out = trimmed_out.mix(TRIMGALORE.out.reads)
        // ---  trimmed_out = trimmed_out.mix(TRIMMOMATIC.out.reads)
        trimgalore_log          = trimgalore_log.mix(TRIMGALORE.out.log)
        trimgalore_unpaired     = trimgalore_unpaired.mix(TRIMGALORE.out.unpaired)
        trimgalore_html         = trimgalore_html.mix(TRIMGALORE.out.html)
        trimgalore_zip          = trimgalore_zip.mix(TRIMGALORE.out.zip)
        ch_versions = ch_versions.mix(TRIMGALORE.out.versions)
        trim_reads = TRIMGALORE.out.reads
    } 
    // Postprocessing QC
    if (!skip_fastqc) {
        PostprocessFASTQC(trim_reads)
        ch_versions = ch_versions.mix(PostprocessFASTQC.out.versions)
        fastqc_html = fastqc_html.mix(PostprocessFASTQC.out.html)
        fastqc_zip = fastqc_zip.mix(PostprocessFASTQC.out.zip)
    }

    emit:
    reads               = trim_reads           // channel: [ val(meta), [ reads ] ]
    fastqc_html         = fastqc_html          // channel: [ val(meta), path(html) ]
    fastqc_zip          = fastqc_zip           // channel: [ val(meta), path(zip) ]
    trimgalore_log                             // channel: [ val(meta), path(log) ]
    trimgalore_unpaired                        // channel: [ val(meta), [ unpaired_reads ] ]
    trimgalore_html                            // channel: [ val(meta), path(html) ]
    trimgalore_zip                             // channel: [ val(meta), path(zip) ]
    versions = ch_versions                     // channel: [ ch_versions.yml ]
}

