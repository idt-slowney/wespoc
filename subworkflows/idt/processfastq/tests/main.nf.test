nextflow_workflow {

    name "Test Subworkflow PROCESSFASTQ"
    script "../main.nf"
    workflow "PROCESSFASTQ"

    tag "subworkflows"
    tag "subworkflows_nfcore"
    tag "subworkflows/processfastq"
    tag "fastqc"
    tag "seqtk/sample"
    tag "cutadapt/extractumi"
    tag "trimgalore"

    test("Paired end fastq input, only skip UMI") {

        when {
            workflow {
                """
                skip_fastqc        = false 
                skip_downsample    = false
                umi_flag           = false
                skip_extractumi    = true
                read_structure     = '8M143T 8M143T'
                skip_trimming      = false

                input[0] = Channel.of([
                    [ id:'test', single_end:false, sample_size:10000 ],
                    [
                        file(params.test_data['homo_sapiens']['illumina']['test_1_fastq_gz'], checkIfExists: true),
                        file(params.test_data['homo_sapiens']['illumina']['test_2_fastq_gz'], checkIfExists: true)
                    ]
                ])
                input[1] = skip_fastqc
                input[2] = skip_downsample
                input[3] = umi_flag
                input[4] = skip_extractumi
                input[5] = read_structure
                input[6] = skip_trimming
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(
                        workflow.out.reads,
                        workflow.out.versions
                    ).match()
                }
            )
        }
    }

    test("Paired end fastq input, skip fastqc") {

        when {
            workflow {
                """
                skip_fastqc        = true 
                skip_downsample    = false
                umi_flag           = false
                skip_extractumi    = true
                read_structure     = '8M143T 8M143T'
                skip_trimming      = false

                input[0] = Channel.of([
                    [ id:'test', single_end:false, sample_size:10000 ],
                    [ file(params.test_data['homo_sapiens']['illumina']['test_1_fastq_gz'], checkIfExists: true),
                      file(params.test_data['homo_sapiens']['illumina']['test_2_fastq_gz'], checkIfExists: true) ]
                ])
                input[1] = skip_fastqc
                input[2] = skip_downsample
                input[3] = umi_flag
                input[4] = skip_extractumi
                input[5] = read_structure
                input[6] = skip_trimming
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success},
                { assert !workflow.out.fastqc_html },
                { assert !workflow.out.fastqc_zip },
                { assert snapshot(
                    workflow.out.reads,
                    workflow.out.versions).match() }
            )
        }
    }

    test("Paired end fastq input with UMI") {

        when {
            workflow {
                """
                skip_fastqc        = false 
                skip_downsample    = false
                umi_flag           = true
                skip_extractumi    = false
                read_structure     = '8M143T 8M143T'
                skip_trimming      = false

                input[0] = Channel.of([
                    [ id:'test', single_end:false, sample_size:10000 ],
                    [ file(params.test_data['homo_sapiens']['illumina']['test_umi_1_fastq_gz'], checkIfExists: true),
                      file(params.test_data['homo_sapiens']['illumina']['test_umi_2_fastq_gz'], checkIfExists: true) ]
                ])
                input[1] = skip_fastqc
                input[2] = skip_downsample
                input[3] = umi_flag
                input[4] = skip_extractumi
                input[5] = read_structure
                input[6] = skip_trimming
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success},
                { assert snapshot(
                    workflow.out.reads,
                    workflow.out.versions).match() }
            )
        }
    }

    test("Skip all") {

        when {
            workflow {
                """
                skip_fastqc        = true 
                skip_downsample    = true
                umi_flag           = true
                skip_extractumi    = true
                read_structure     = '8M143T 8M143T'
                skip_trimming      = true

                input[0] = Channel.of([
                    [ id:'test', single_end:false, sample_size:10000 ],
                    [ file(params.test_data['homo_sapiens']['illumina']['test_umi_1_fastq_gz'], checkIfExists: true),
                      file(params.test_data['homo_sapiens']['illumina']['test_umi_2_fastq_gz'], checkIfExists: true) ]
                ])
                input[1] = skip_fastqc
                input[2] = skip_downsample
                input[3] = umi_flag
                input[4] = skip_extractumi
                input[5] = read_structure
                input[6] = skip_trimming
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success},
                { assert !workflow.out.fastqc_html },
                { assert !workflow.out.fastqc_zip },
                { assert !workflow.out.trimgalore_log },
                { assert snapshot(
                    workflow.out.reads,
                    workflow.out.versions).match() }
            )
        }
    }
}

