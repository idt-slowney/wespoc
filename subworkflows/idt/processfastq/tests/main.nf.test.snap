{
    "Skip all": {
        "content": [
            [
                [
                    {
                        "id": "test",
                        "single_end": false,
                        "sample_size": 10000
                    },
                    [
                        "/nf-core/test-datasets/modules/data/genomics/homo_sapiens/illumina/fastq/test.umi_1.fastq.gz",
                        "/nf-core/test-datasets/modules/data/genomics/homo_sapiens/illumina/fastq/test.umi_2.fastq.gz"
                    ]
                ]
            ],
            [
                
            ]
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.3"
        },
        "timestamp": "2024-08-09T13:36:17.861110728"
    },
    "Paired end fastq input with UMI": {
        "content": [
            [
                [
                    {
                        "id": "test",
                        "single_end": false,
                        "sample_size": 10000
                    },
                    [
                        "test_1_val_1.fq.gz:md5,4adc9f02b4f8b503339966b27effe711",
                        "test_2_val_2.fq.gz:md5,dfc08612c22fb4993b4263f97a07a7c0"
                    ]
                ]
            ],
            [
                "versions.yml:md5,06e5e3503bcf65395062c31d07a590f3",
                "versions.yml:md5,77e3d93e13faa7e33fcad5bc8f577f2b",
                "versions.yml:md5,7b9f999b76de4a7d95267dfd2d9ddec0",
                "versions.yml:md5,8bdfbd4b635d2953ad6d090dfe5494c8",
                "versions.yml:md5,e0d4c6ee367a4d6600b7268278549bae"
            ]
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.3"
        },
        "timestamp": "2024-08-09T13:36:03.665176057"
    },
    "Paired end fastq input, only skip UMI": {
        "content": [
            [
                [
                    {
                        "id": "test",
                        "single_end": false,
                        "sample_size": 10000
                    },
                    [
                        "test_1_val_1.fq.gz:md5,1155b07c8311d64495821fb45014e03e",
                        "test_2_val_2.fq.gz:md5,6c55135fa6eb1138d0c479e8f7dc5fae"
                    ]
                ]
            ],
            [
                "versions.yml:md5,06e5e3503bcf65395062c31d07a590f3",
                "versions.yml:md5,77e3d93e13faa7e33fcad5bc8f577f2b",
                "versions.yml:md5,7b9f999b76de4a7d95267dfd2d9ddec0",
                "versions.yml:md5,e0d4c6ee367a4d6600b7268278549bae"
            ]
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.3"
        },
        "timestamp": "2024-08-09T13:27:58.290125093"
    },
    "Paired end fastq input, skip fastqc": {
        "content": [
            [
                [
                    {
                        "id": "test",
                        "single_end": false,
                        "sample_size": 10000
                    },
                    [
                        "test_1_val_1.fq.gz:md5,1155b07c8311d64495821fb45014e03e",
                        "test_2_val_2.fq.gz:md5,6c55135fa6eb1138d0c479e8f7dc5fae"
                    ]
                ]
            ],
            [
                "versions.yml:md5,77e3d93e13faa7e33fcad5bc8f577f2b",
                "versions.yml:md5,e0d4c6ee367a4d6600b7268278549bae"
            ]
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.3"
        },
        "timestamp": "2024-08-09T13:31:03.351857852"
    }
}